src/lib/system_logger.ts

```diff
@@ -0,0 +1,115 @@
+const systemLogTag = '__nfSystemLog'
+
+const serializeError = (error: Error): Record<string, unknown> => {
+  const cause = error?.cause instanceof Error ? serializeError(error.cause) : error.cause
+
+  return {
+    error: error.message,
+    error_cause: cause,
+    error_stack: error.stack,
+  }
+}
+
+// eslint pretends there's a different enum at the same place - it's wrong!
+// eslint-disable-next-line no-shadow
+export enum LogLevel {
+  Debug = 1,
+  Info,
+  Warn,
+  Error,
+}
+
+class SystemLogger {
+  // eslint-disable-next-line no-useless-constructor
+  constructor(
+    private readonly fields: Record<string, unknown> = {},
+    private readonly logLevel = LogLevel.Info,
+    private readonly samplingRate = 1,
+  ) {}
+
+  private doLog(logger: typeof console.log, message: string) {
+    if (this.samplingRate < 1 && Math.random() > this.samplingRate) {
```

Comment:
I've added support for sampling here, because I feel like it's a useful thing for collecting metrics without increasing the log volume too much. The more I think about it, though, the more I feel like the sampling rate should be set on a per-log basis, instead of on a per-logger basis.  @eduardoboucas curious what you think -  is it useful to have sampling built-in to the logger? If yes, should we also add it to the edge functions logger?